File: tools/brace.py
"""Give every if/else/for/while/do body a brace block.

The point is not style.  Everything after this needs to know where a body
begins and ends; make it syntactic and the tools stop parsing C, leave it
implicit and every later pass is a fresh chance to get an extent wrong.

Runs after joinparens.py and splitheads.py, so a head is a whole line ending in
`)` (or the word `else` or `do`) and a body starts on the next line.

Traps this handles:
  * `} while (cond);` is the tail of a do-while, not a head -- a head line never
    ends in `;`.
  * `else if (...)` is one statement: the `if` is the head, not the body.
  * a `do` body's terminating `while (...)` stays outside the braces.
  * a statement can span lines when it contains an initialiser list, so the end
    is "brace depth back to zero and the line ends in `;`", not "the next `;`".

Usage: brace.py <file>
"""
import re
import sys

sys.path.insert(0, __file__.rsplit('/', 1)[0])
import cutil

HEAD = re.compile(r'^(\s*)(?:\}\s*)?(if|for|while|switch)\s*\(')
ELSE_IF = re.compile(r'^(\s*)else\s+if\s*\(')
ELSE = re.compile(r'^(\s*)else\s*$')
DO = re.compile(r'^(\s*)do\s*$')


def is_skippable(line):
    s = line.strip()
    return not s or s.startswith('#') or s.startswith('//')


def next_code(lines, i):
    while i < len(lines) and is_skippable(lines[i]):
        i += 1
    return i


def head_of(line, index=None, do_terms=()):
    """(indent, kind) if this line is a control head with a body, else None.

    A `while` that terminates a do-while is not a head.  Usually it says so by
    ending in `;`, but it can also be written

        }
        while (cond)
            ;

    with the semicolon on its own line, and then it is indistinguishable from a
    real while-head by looking at the line alone.  Those are found in a first
    pass and passed in as do_terms.
    """
    s = line.rstrip()
    if is_skippable(s):
        return None
    if index is not None and index in do_terms:
        return None
    body = s.strip()
    if body.endswith(';'):
        return None                      # `while (cond);` closing a do
    m = DO.match(s)
    if m:
        return (m.group(1), 'do')
    m = ELSE.match(s)
    if m:
        return (m.group(1), 'else')
    m = ELSE_IF.match(s)
    if m and s.endswith(')'):
        return (m.group(1), 'if')
    m = HEAD.match(s)
    if m and s.endswith(')'):
        # Only an `if` may swallow a following `else`.  A for/while/switch that
        # happens to be an if's body must not: doing so puts the closing brace
        # on the far side of the else and moves the else's body inside the loop.
        return (m.group(1), 'if' if m.group(2) == 'if' else 'loop')
    return None


def stmt_end(lines, i, do_terms=()):
    """Index of the last line of the statement starting at line i."""
    i = next_code(lines, i)
    if i >= len(lines):
        return i
    h = head_of(lines[i], i, do_terms)
    if h is not None:
        kind = h[1]
        j = next_code(lines, i + 1)
        e = stmt_end(lines, j, do_terms)
        if kind == 'do':
            k = next_code(lines, e + 1)
            if k < len(lines) and re.match(r'^\s*(\}\s*)?while\b', lines[k]):
                # The do-while ends at its semicolon, which is not always on
                # the same line as the `while`.
                while k < len(lines) and not lines[k].rstrip().endswith(';'):
                    k += 1
                return k
            return e
        if kind == 'if':
            k = next_code(lines, e + 1)
            if k < len(lines) and re.match(r'^\s*(\}\s*)?else\b', lines[k]):
                return stmt_end(lines, k, do_terms)
        return e
    # a plain statement, or a brace block
    depth = 0
    started = False
    j = i
    while j < len(lines):
        if is_skippable(lines[j]):
            j += 1
            continue
        b = cutil.blank(lines[j])
        depth += b.count('{') - b.count('}')
        if '{' in b:
            started = True
        s = lines[j].rstrip()
        if depth <= 0 and (s.endswith(';') or (started and s.endswith('}'))):
            return j
        if depth <= 0 and s.endswith(':'):
            return j
        j += 1
    return len(lines) - 1


def main():
    path = sys.argv[1]
    lines = open(path, encoding='utf-8', errors='surrogateescape').read().split('\n')
    # Find the do-terminating `while` lines first, to a fixpoint: identifying
    # one can change the extent of an enclosing statement, which can reveal
    # another.
    do_terms = set()
    for _ in range(5):
        found = set()
        for i, line in enumerate(lines):
            h = head_of(line, i, do_terms)
            if h is None or h[1] != 'do':
                continue
            j = next_code(lines, i + 1)
            e = stmt_end(lines, j, do_terms)
            k = next_code(lines, e + 1)
            if k < len(lines) and re.match(r'^\s*(\}\s*)?while\b', lines[k]):
                found.add(k)
        if found == do_terms:
            break
        do_terms |= found
    print('%d do-terminating while lines identified' % len(do_terms))

    opens, closes = {}, {}
    braced = 0
    for i, line in enumerate(lines):
        h = head_of(line, i, do_terms)
        if h is None:
            continue
        indent = h[0]
        j = next_code(lines, i + 1)
        if j >= len(lines):
            continue
        if lines[j].strip().startswith('{'):
            continue
        e = stmt_end(lines, j, do_terms)
        opens.setdefault(i, []).append(indent)
        closes.setdefault(e, []).append(indent)
        braced += 1
    out = []
    for i, line in enumerate(lines):
        out.append(line)
        for indent in opens.get(i, []):
            out.append(indent + '{')
        for indent in reversed(closes.get(i, [])):
            out.append(indent + '}')
    # a close for line e and an open for line e cannot both be right; the loop
    # above emits opens first, which is the order they nest in.
    text = '\n'.join(out)
    if not text.endswith('\n'):
        text += '\n'
    open(path, 'w', encoding='utf-8', errors='surrogateescape').write(text)
    print('%d bodies braced; %d lines -> %d' % (braced, len(lines), len(out)))


if __name__ == '__main__':
    main()

File: tools/go/internal/canon/brace.go
package canon

import (
	"bytes"

	"slimvim.local/tools/internal/cutil"
)

// isSkippable is brace.py's: blank, a directive, or a // comment.  A body is
// none of those, so the scan steps over them to find the statement a head
// governs -- `if (x)` followed by `#ifdef` followed by `y();` governs y().
func isSkippable(line []byte) bool {
	s := trim(line)
	return len(s) == 0 || s[0] == '#' || bytes.HasPrefix(s, []byte("//"))
}

func nextCode(lines [][]byte, i int) int {
	for i < len(lines) && isSkippable(lines[i]) {
		i++
	}
	return i
}

// skipCloseBrace consumes an optional `}` and the whitespace after it, which
// is the `(?:\}\s*)?` several of brace.py's patterns open with.
func skipCloseBrace(s []byte) []byte {
	if len(s) > 0 && s[0] == '}' {
		s = s[1:]
		for len(s) > 0 && isSpace(s[0]) {
			s = s[1:]
		}
	}
	return s
}

// wordThenParen matches one of the keywords, optional space, then '(' .
func wordThenParen(s []byte, kws ...string) (string, bool) {
	for _, kw := range kws {
		if !bytes.HasPrefix(s, []byte(kw)) {
			continue
		}
		r := s[len(kw):]
		if len(r) > 0 && isWord(r[0]) {
			continue
		}
		for len(r) > 0 && isSpace(r[0]) {
			r = r[1:]
		}
		if len(r) > 0 && r[0] == '(' {
			return kw, true
		}
	}
	return "", false
}

// onlyWord matches `^(\s*)kw\s*$`.
func onlyWord(s []byte, kw string) bool {
	r := ltrim(s)
	if !bytes.HasPrefix(r, []byte(kw)) {
		return false
	}
	return len(trim(r[len(kw):])) == 0
}

// elseIfHead matches `^(\s*)else\s+if\s*\(` -- note the REQUIRED whitespace
// between else and if, which is why it is not wordThenParen("else", "if").
func elseIfHead(s []byte) bool {
	r := ltrim(s)
	if !bytes.HasPrefix(r, []byte("else")) {
		return false
	}
	r = r[len("else"):]
	if len(r) == 0 || !isSpace(r[0]) {
		return false
	}
	r = ltrim(r)
	_, ok := wordThenParen(r, "if")
	return ok
}

// startsWithCloseThenWord matches `^\s*(\}\s*)?kw\b`.
func startsWithCloseThenWord(s []byte, kw string) bool {
	return startsWithWord(skipCloseBrace(ltrim(s)), kw)
}

type head struct {
	indent []byte
	kind   string // "do", "else", "if" or "loop"
}

// headOf returns the head this line is, or ok=false.
//
// A `while` that terminates a do-while is not a head.  Usually it says so by
// ending in ';', but it can also be written with the semicolon on a line of
// its own, and then it cannot be told from a real while-head by looking at the
// line alone.  Those are found in a first pass and passed in as doTerms.
func headOf(line []byte, index int, doTerms map[int]bool) (head, bool) {
	s := rtrim(line)
	if isSkippable(s) {
		return head{}, false
	}
	if index >= 0 && doTerms[index] {
		return head{}, false
	}
	if bytes.HasSuffix(trim(s), []byte(";")) {
		return head{}, false // `while (cond);` closing a do
	}
	indent := indentOf(s)
	if onlyWord(s, "do") {
		return head{indent, "do"}, true
	}
	if onlyWord(s, "else") {
		return head{indent, "else"}, true
	}
	if elseIfHead(s) && bytes.HasSuffix(s, []byte(")")) {
		return head{indent, "if"}, true
	}
	if kw, ok := wordThenParen(skipCloseBrace(ltrim(s)), "if", "for", "while", "switch"); ok &&
		bytes.HasSuffix(s, []byte(")")) {
		// Only an `if` may swallow a following `else`.  A for/while/switch
		// that happens to be an if's body must not: doing so would put the
		// closing brace on the far side of the else and move the else's body
		// inside the loop.
		if kw == "if" {
			return head{indent, "if"}, true
		}
		return head{indent, "loop"}, true
	}
	return head{}, false
}

// stmtEnd returns the index of the last line of the statement starting at i.
func stmtEnd(lines [][]byte, i int, doTerms map[int]bool) int {
	i = nextCode(lines, i)
	if i >= len(lines) {
		return i
	}
	if h, ok := headOf(lines[i], i, doTerms); ok {
		j := nextCode(lines, i+1)
		e := stmtEnd(lines, j, doTerms)
		switch h.kind {
		case "do":
			k := nextCode(lines, e+1)
			if k < len(lines) && startsWithCloseThenWord(lines[k], "while") {
				// The do-while ends at its semicolon, which is not always on
				// the same line as the `while`.
				for k < len(lines) && !bytes.HasSuffix(rtrim(lines[k]), []byte(";")) {
					k++
				}
				return k
			}
			return e
		case "if":
			k := nextCode(lines, e+1)
			if k < len(lines) && startsWithCloseThenWord(lines[k], "else") {
				return stmtEnd(lines, k, doTerms)
			}
		}
		return e
	}

	// A plain statement, or a brace block.  The end is "brace depth back to
	// zero and the line ends in ';'", not "the next ';'", because a statement
	// can span lines when it contains an initialiser list.
	depth := 0
	started := false
	for j := i; j < len(lines); j++ {
		if isSkippable(lines[j]) {
			continue
		}
		b := cutil.Blank(lines[j])
		depth += bytes.Count(b, []byte("{")) - bytes.Count(b, []byte("}"))
		if bytes.IndexByte(b, '{') >= 0 {
			started = true
		}
		s := rtrim(lines[j])
		if depth <= 0 && (bytes.HasSuffix(s, []byte(";")) ||
			(started && bytes.HasSuffix(s, []byte("}")))) {
			return j
		}
		if depth <= 0 && bytes.HasSuffix(s, []byte(":")) {
			return j
		}
	}
	return len(lines) - 1
}

// Brace gives every if/else/for/while/do body a brace block.
//
// The point is not style.  Everything after this needs to know where a body
// begins and ends: make it syntactic and the tools stop parsing C, leave it
// implicit and every later pass is a fresh chance to get an extent wrong.
//
// It runs after JoinParens and SplitHeads, so a head is a whole line ending in
// ')' -- or the word `else` or `do` -- and a body starts on the next line.
//
// It returns the do-terminating count as well, because brace.py prints TWO
// lines where every other tool prints one, and tools/sweep.sh reads only the
// last.  Both are printed here: matching a drop-in means matching what it
// writes, not only what its caller happens to read.
func Brace(src []byte) (out []byte, doTermCount, braced, nIn, nOut int) {
	lines := split(src)
	nIn = len(lines)

	// The do-terminating `while` lines, to a fixpoint: identifying one can
	// change the extent of an enclosing statement, which can reveal another.
	doTerms := map[int]bool{}
	for round := 0; round < 5; round++ {
		found := map[int]bool{}
		for i, line := range lines {
			h, ok := headOf(line, i, doTerms)
			if !ok || h.kind != "do" {
				continue
			}
			j := nextCode(lines, i+1)
			e := stmtEnd(lines, j, doTerms)
			k := nextCode(lines, e+1)
			if k < len(lines) && startsWithCloseThenWord(lines[k], "while") {
				found[k] = true
			}
		}
		if sameSet(found, doTerms) {
			break
		}
		for k := range found {
			doTerms[k] = true
		}
	}
	doTermCount = len(doTerms)

	opens := map[int][][]byte{}
	closes := map[int][][]byte{}
	for i, line := range lines {
		h, ok := headOf(line, i, doTerms)
		if !ok {
			continue
		}
		j := nextCode(lines, i+1)
		if j >= len(lines) {
			continue
		}
		if bytes.HasPrefix(trim(lines[j]), []byte("{")) {
			continue
		}
		e := stmtEnd(lines, j, doTerms)
		opens[i] = append(opens[i], h.indent)
		closes[e] = append(closes[e], h.indent)
		braced++
	}

	kept := make([][]byte, 0, len(lines)+2*braced)
	for i, line := range lines {
		kept = append(kept, line)
		for _, indent := range opens[i] {
			kept = append(kept, concat(indent, []byte("{")))
		}
		// A close for line e and an open for line e cannot both be right; the
		// opens are emitted first, which is the order they nest in.
		cl := closes[i]
		for k := len(cl) - 1; k >= 0; k-- {
			kept = append(kept, concat(cl[k], []byte("}")))
		}
	}

	nOut = len(kept)
	return joinWithFinalNewline(kept), doTermCount, braced, nIn, nOut
}

func sameSet(a, b map[int]bool) bool {
	if len(a) != len(b) {
		return false
	}
	for k := range a {
		if !b[k] {
			return false
		}
	}
	return true
}
